runtime/src/iree-amd-aie/driver/xrt-lite/device.cc

```diff
@@ -0,0 +1,296 @@
+    iree_device_size_t allocation_size,
+    iree_hal_buffer_t** IREE_RESTRICT out_buffer) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_lite_device* device =
+      reinterpret_cast<iree_hal_xrt_lite_device*>(base_value);
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_semaphore_list_wait(wait_semaphore_list,
+                                       iree_infinite_timeout()));
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_allocator_allocate_buffer(device->device_allocator, params,
+                                             allocation_size, out_buffer));
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_semaphore_list_signal(signal_semaphore_list));
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
+}
+
+static iree_string_view_t iree_hal_xrt_lite_device_id(
+    iree_hal_device_t* base_value) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_lite_device* device =
+      reinterpret_cast<iree_hal_xrt_lite_device*>(base_value);
+
+  IREE_TRACE_ZONE_END(z0);
+  return device->identifier;
+}
+
+static void iree_hal_xrt_lite_device_destroy(iree_hal_device_t* base_value) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_lite_device* device =
+      reinterpret_cast<iree_hal_xrt_lite_device*>(base_value);
+
+  iree_hal_allocator_release(device->device_allocator);
+  delete device->shim_device;
+  iree_allocator_free(device->host_allocator, device);
+
+  IREE_TRACE_ZONE_END(z0);
+};
+
+static iree_allocator_t iree_hal_xrt_lite_device_host_allocator(
+    iree_hal_device_t* base_value) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_lite_device* device =
+      reinterpret_cast<iree_hal_xrt_lite_device*>(base_value);
+
+  IREE_TRACE_ZONE_END(z0);
+  return device->host_allocator;
+}
+
+static iree_hal_allocator_t* iree_hal_xrt_lite_device_device_allocator(
+    iree_hal_device_t* base_value) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_lite_device* device =
+      reinterpret_cast<iree_hal_xrt_lite_device*>(base_value);
+
+  IREE_TRACE_ZONE_END(z0);
+  return device->device_allocator;
+}
+
+void iree_hal_xrt_lite_device_options_initialize(
+    iree_hal_xrt_lite_device_options* out_options) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  memset(out_options, 0, sizeof(*out_options));
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+iree_status_t iree_hal_xrt_lite_device_create(
+    iree_string_view_t identifier,
+    const iree_hal_xrt_lite_device_options* options,
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device) {
+  IREE_ASSERT_ARGUMENT(options);
+  IREE_ASSERT_ARGUMENT(out_device);
+  *out_device = nullptr;
+
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_lite_device* device = nullptr;
+  iree_host_size_t total_size = iree_sizeof_struct(*device) + identifier.size;
+  IREE_RETURN_IF_ERROR(
+      iree_allocator_malloc(host_allocator, total_size, (void**)&device));
+  device = new (device) iree_hal_xrt_lite_device(options, host_allocator);
```

Comment:
can you add a comment that this is calling the constructor of `iree_hal_xrt_lite_device` . Took me a while to figure out as we normally dont see structs with contructors in a hal implementation.

Reply:
If enough people find this super weird I'll just make it conventional. Vote here if you don't like it!

Reply:
The placement new doesn't really bother me here, what does trip me up is the dangling memory after the new-ed block that we fill with the identifier. This is what we do in C but feels weirder in C++.

Reply:
any alternative come to mind? in fact in this case (`iree_hal_xrt_lite_device`) i actually have `iree_string_view_t identifier;` in the wrong order/position.